crypten/mpc/primitives/circuit.py

```diff
@@ -55,14 +55,15 @@ def __SPK_circuit(S, P):
     # fmt: off
     # Tree reduction circuit
     for i in range(__LOG_BITS):
-        in_mask = __MASKS[i]                # Start of arrows
-        out_mask = __OUT_MASKS[i]           # End of arrows
-        not_out_mask = out_mask ^ -1        # Not (end of arrows)
+        in_mask = __MASKS[i].to(SP.device)                   # Start of arrows
+        out_mask = __OUT_MASKS[i].to(SP[1].device)           # End of arrows
+        not_out_mask = out_mask ^ -1                         # Not (end of arrows)
+        multiplier = __MULTIPLIERS[i].to(SP.device)
```

Comment:
We call this function a lot. Rather than casting constants to device each time, it may be useful to have these cached on each device.

Reply:
That sounds reasonable to me. I've modified `circuit.py` a little so that `__MASKS`, `__OUT_MASKS`, and `__MULTIPLIERS` are generated by a new function `__SPK_circuit_constants`, where the return values are cached using `functools.lru_cache`. Let me know if you have any issues with the new implementation.